aten/src/ATen/native/cuda/DistributionTemplates.h

```diff
@@ -557,45 +536,16 @@ struct GeometricKernel {
 
 template<typename RNG>
 void exponential_kernel(TensorIterator& iter, double lambda_, RNG gen) {
-  // Note that HIP doesn't support std::nextafter in device code.
-  auto nextafter_1_0_float = std::nextafter(1.0f, 0.0f);
-  auto nextafter_1_0_double = std::nextafter(1.0, 0.0);
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "exponential_cuda", [&] {
     using accscalar_t = at::acc_type<scalar_t, true>;
     auto lambda = static_cast<accscalar_t>(lambda_);
-    if (std::is_same<scalar_t, double>::value) {
-      // define lambda for exponential transformation
-      auto exponential_func = [lambda, nextafter_1_0_double] __device__ (accscalar_t rand) {
-        accscalar_t sample;
-        // curand_uniform has (0,1] bounds. log(1) is 0 and exponential excludes 0.
-        // Hence, squash the 1 to just below 1.
-        if(rand == static_cast<accscalar_t>(1.0)) {
-          sample = ::log(nextafter_1_0_double);
-        } else {
-          sample = ::log(rand);
-        }
-        return static_cast<scalar_t>(static_cast<accscalar_t>(-1.0) / lambda * sample);
-      };
-      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls/2>(iter,
-        gen,
-        [] __device__ (curandStatePhilox4_32_10_t* state) { return curand_uniform2_double(state); },
-        exponential_func);
-    } else {
-      // use __logf fast approximation for peak bandwidth
-      auto exponential_func = [lambda, nextafter_1_0_float] __device__ (accscalar_t rand) {
-        accscalar_t sample;
-        if(rand == static_cast<accscalar_t>(1.0)) {
-          sample = __logf(nextafter_1_0_float);
```

Comment:
It seems to me that you have completely lost the `nextafter_1_0_float` logic here; the rewritten version calls `at::log` but there is no test for when you try to take `log` of `1.0`. Is there an argument why this logic is no longer necessary? I didn't see this remarked upon in the PR description.

Reply:
1) Yes, I dropped this "nextafter" logic that doesn't make any sense for me.
2) All this "nextafter" seems weird for me and I unified it with prior `uniform_kernel`'s approach: `auto reverse_bound_rand = rand == static_cast<accscalar_t>(1.0) ? static_cast<accscalar_t>(0.0) : rand;`
3) There were no tests before and we just trusted this code because of line `if(rand == static_cast<accscalar_t>(1.0)) ...` and I replaced if with `auto reverse_bound_rand = rand == static_cast<accscalar_t>(1.0) ? static_cast<accscalar_t>(0.0) : rand;`

So, the logic is:
1) run curand uniform that may return 1
2) if-1-then-0
3) run `transformation::exponential` that calls `log`

The only way I can test this code is to move `if-1-then-0` to `transformation::exponential` and test it in a unit test, but then this if statement would run always not only when we use curand and might affect cpu's `exponential_kernel` performance.

Reply:
> it doesn't make any sense for me

I don't understand the logic either 😄, but I do believe I understand the goal. I believe the end goal is to ensure exponential random variables are distributed in the open half line `(0,infinity)`, in particular there is a lot of probabilistic code out there that assumes an exponential variate is never equal to zero. The means to achieve this goal was to use nextafter to clamp a value in the half open interval `(0,1]` to a value in the open interval `(0,1)`. Then later on log transformation that `(0,1)` value would be transformed to a `(0,infinity)` value.

> The only way I can test this code ...

Good point: these edge effects are difficult to test. Zooming out, do you think it would be possible to insert a mock random engine that flipped a coin and retuned either the min possible value (something like finfo.tiny or finfo.min) or the max value (here 1.0)? We could alternatively have a pair (min_mock, max_mock), but some random generators consume multiple values so it would be good to test edge effects like `my_transform(min_value, min_value, max_value)`. That seems like a decent strategy to test all these nuanced edge cases that require such detailed code review.

Reply:
Actually, this was the piece in the PR description that bothered me most. 
```
run curand uniform that may return 1
if-1-then-0
```
produces 0 with ~ 2^-24 probability - which would be kinda fine, if the next positive number was 2^-24, and it was produced with the same probability. But the next positive number is 2^-32, and it is produced with the probability of 2^-32, so we are severely oversampling 0-s.

Reply:
Would this be an item to open as a separate PR?